src/filesystem/nodes/metadata/mime_type.rs

```diff
@@ -0,0 +1,252 @@
+use crate::prelude::nodes::NodeName;
+
+#[derive(Default)]
+pub struct MimeGuesser {
+    name: Option<String>,
+    data: Vec<u8>,
+}
+
+impl MimeGuesser {
+    const MP3_RATES: [u32; 15] = [
+        0, 32000, 40000, 48000, 56000, 64000, 80000, 96000, 112000, 128000, 160000, 192000, 224000,
+        256000, 320000,
+    ];
+
+    const MP25_RATES: [u32; 15] = [
+        0, 8000, 16000, 24000, 32000, 40000, 48000, 56000, 64000, 80000, 96000, 112000, 128000,
+        144000, 160000,
+    ];
+
+    const SAMPLE_RATES: [u32; 3] = [44100, 48000, 32000];
+
+    pub fn with_name(mut self, name: NodeName) -> Self {
+        match name {
+            NodeName::Named(name) => self.name = Some(name.clone()),
+            NodeName::Root => {}
+        }
+        self
+    }
+
+    pub fn with_data(mut self, data: &[u8]) -> Self {
+        self.data.extend_from_slice(data);
+        self
+    }
+
+    pub fn guess_mime_type(&self) -> Option<mime::MediaType> {
+        self.pattern_match()
+            .or_else(|| self.algorithm_match())
+            .or_else(|| self.extension_match())
+    }
+
+    fn extension_match(&self) -> Option<mime::MediaType> {
+        if let Some(name) = self.name.as_ref() {
+            let guess = mime_guess::from_path(name);
+            if !guess.is_empty() {
+                return mime::MediaType::parse(guess.first()?.as_ref()).ok();
+            }
+        }
+        None
+    }
+
+    fn pattern_match(&self) -> Option<mime::MediaType> {
```

Comment:
Dang I wasn't expecting to see an actual byte based mime guesser in here that's pretty neat. I've never thought of matching variable length byte arrays like this, that's pretty neat and probably pretty efficient.